dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/trackedentity/AbstractTrackedEntityInstanceService.java

```diff
@@ -390,19 +390,27 @@ else if ( !FeatureType.NONE.equals( dtoEntityInstance.getFeatureType() ) && dtoE
     public ImportSummaries addTrackedEntityInstances( List<TrackedEntityInstance> trackedEntityInstances,
         ImportOptions importOptions )
     {
-        List<List<TrackedEntityInstance>> partitions = Lists.partition( trackedEntityInstances, FLUSH_FREQUENCY );
         importOptions = updateImportOptions( importOptions );
         ImportSummaries importSummaries = new ImportSummaries();
         List<Enrollment> enrollments = new ArrayList<>();
 
+        List<String> conflictingTeiUids = checkForExistingTeisIncludingDeleted( trackedEntityInstances, importSummaries );
+
+        List<TrackedEntityInstance> validTeis = trackedEntityInstances.stream()
+            .filter( tei -> !conflictingTeiUids.contains( tei.getTrackedEntityInstance() ) )
+            .collect( Collectors.toList() );
+
+        List<List<TrackedEntityInstance>> partitions = Lists.partition( validTeis, FLUSH_FREQUENCY );
+
```

Comment:
For increased readability, I would move the entire "validity" check to its own function:

```
/**
     * Filters out Tracked Entity Instances which are already present in the database (regardless of the 'deleted' state)
     *  
     * 
     * @param trackedEntityInstances
     * @param importSummaries
     * @return
     */
    private List<TrackedEntityInstance> resolveImportableTeis(List<TrackedEntityInstance> trackedEntityInstances, ImportSummaries importSummaries) {

        List<String> conflictingTeiUids = checkForExistingTeisIncludingDeleted( trackedEntityInstances, importSummaries );

        return trackedEntityInstances.stream()
                .filter( tei -> !conflictingTeiUids.contains( tei.getTrackedEntityInstance() ) )
                .collect( Collectors.toList() );
    }
```
and then just call: 

```
List<TrackedEntityInstance> validTeis = resolveImportableTeis( trackedEntityInstances, importSummaries);
```

Reply:
Yep. I moved it into a separate method.